src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs

```diff
@@ -59,25 +59,48 @@ protected override void OnNext(string eventName, object arg)
 
             if (lastChar == 't')
             {
-                if (eventName.AsSpan().Slice(PrefixLength).SequenceEqual("Hosting.HttpRequestIn.Start"))
+                if (ReferenceEquals(eventName, hostingHttpRequestInStartEventKey))
```

Comment:
What events did you use for the benchmark? You're likely making it worse for the non-matching events, which is the most common case. 

The sequence of events for a typical request is:
```
"Microsoft.AspNetCore.Hosting.HttpRequestIn.Start",
"Microsoft.AspNetCore.Hosting.BeginRequest",
"Microsoft.AspNetCore.Routing.EndpointMatched",
"Microsoft.AspNetCore.Mvc.BeforeAction",
"Microsoft.AspNetCore.Mvc.BeforeOnResourceExecuting",
"Microsoft.AspNetCore.Mvc.AfterOnResourceExecuting",
"Microsoft.AspNetCore.Mvc.BeforeOnActionExecution",
"Microsoft.AspNetCore.Mvc.BeforeOnActionExecuting",
"Microsoft.AspNetCore.Mvc.AfterOnActionExecuting",
"Microsoft.AspNetCore.Mvc.BeforeActionMethod",
"Microsoft.AspNetCore.Mvc.BeforeControllerActionMethod",
"Microsoft.AspNetCore.Mvc.AfterControllerActionMethod",
"Microsoft.AspNetCore.Mvc.AfterActionMethod",
"Microsoft.AspNetCore.Mvc.BeforeOnActionExecuted",
"Microsoft.AspNetCore.Mvc.AfterOnActionExecuted",
"Microsoft.AspNetCore.Mvc.AfterOnActionExecution",
"Microsoft.AspNetCore.Mvc.BeforeOnResultExecuting",
"Microsoft.AspNetCore.Mvc.AfterOnResultExecuting",
"Microsoft.AspNetCore.Mvc.BeforeActionResult",
"Microsoft.AspNetCore.Mvc.ViewFound",
"Microsoft.AspNetCore.Mvc.BeforeView",
"Microsoft.AspNetCore.Mvc.Razor.BeforeViewPage",
"Microsoft.AspNetCore.Mvc.Razor.AfterViewPage",
"Microsoft.AspNetCore.Mvc.AfterView",
"Microsoft.AspNetCore.Mvc.AfterActionResult",
"Microsoft.AspNetCore.Mvc.BeforeOnResultExecuted",
"Microsoft.AspNetCore.Mvc.AfterOnResultExecuted",
"Microsoft.AspNetCore.Mvc.BeforeOnResourceExecuted",
"Microsoft.AspNetCore.Mvc.AfterOnResourceExecuted",
"Microsoft.AspNetCore.Mvc.AfterAction",
"Microsoft.AspNetCore.Hosting.EndRequest",
"Microsoft.AspNetCore.Hosting.HttpRequestIn.Stop"
```

Reply:
the same of the test, thanks for this list, I'll change the benchmark and test it.

Reply:
So I've changed the benchmark with this:

```csharp
        public string[] EventKeys()
        {
            return new string[]
            {
                "Microsoft.AspNetCore.Hosting.HttpRequestIn.Start",
                "Microsoft.AspNetCore.Hosting.BeginRequest",
                "Microsoft.AspNetCore.Routing.EndpointMatched",
                "Microsoft.AspNetCore.Mvc.BeforeAction",
                "Microsoft.AspNetCore.Mvc.BeforeOnResourceExecuting",
                "Microsoft.AspNetCore.Mvc.AfterOnResourceExecuting",
                "Microsoft.AspNetCore.Mvc.BeforeOnActionExecution",
                "Microsoft.AspNetCore.Mvc.BeforeOnActionExecuting",
                "Microsoft.AspNetCore.Mvc.AfterOnActionExecuting",
                "Microsoft.AspNetCore.Mvc.BeforeActionMethod",
                "Microsoft.AspNetCore.Mvc.BeforeControllerActionMethod",
                "Microsoft.AspNetCore.Mvc.AfterControllerActionMethod",
                "Microsoft.AspNetCore.Mvc.AfterActionMethod",
                "Microsoft.AspNetCore.Mvc.BeforeOnActionExecuted",
                "Microsoft.AspNetCore.Mvc.AfterOnActionExecuted",
                "Microsoft.AspNetCore.Mvc.AfterOnActionExecution",
                "Microsoft.AspNetCore.Mvc.BeforeOnResultExecuting",
                "Microsoft.AspNetCore.Mvc.AfterOnResultExecuting",
                "Microsoft.AspNetCore.Mvc.BeforeActionResult",
                "Microsoft.AspNetCore.Mvc.ViewFound",
                "Microsoft.AspNetCore.Mvc.BeforeView",
                "Microsoft.AspNetCore.Mvc.Razor.BeforeViewPage",
                "Microsoft.AspNetCore.Mvc.Razor.AfterViewPage",
                "Microsoft.AspNetCore.Mvc.AfterView",
                "Microsoft.AspNetCore.Mvc.AfterActionResult",
                "Microsoft.AspNetCore.Mvc.BeforeOnResultExecuted",
                "Microsoft.AspNetCore.Mvc.AfterOnResultExecuted",
                "Microsoft.AspNetCore.Mvc.BeforeOnResourceExecuted",
                "Microsoft.AspNetCore.Mvc.AfterOnResourceExecuted",
                "Microsoft.AspNetCore.Mvc.AfterAction",
                "Microsoft.AspNetCore.Hosting.EndRequest",
                "Microsoft.AspNetCore.Hosting.HttpRequestIn.Stop"
            };
        }

        [Benchmark]
        public void CurrentOnNext()
        {
            foreach (var eventName in EventKeys())
                OnNext(eventName, null);
        }

        [Benchmark]
        public void CurrentOnNextNew()
        {
            foreach (var eventName in EventKeys())
                OnNextNew(eventName, null);
        }
```

And the results:

```ini
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19041.630 (2004/?/20H1)
Intel Core i7-1068NG7 CPU 2.30GHz, 1 CPU, 2 logical and 2 physical cores
.NET Core SDK=5.0.100
  [Host]     : .NET Core 3.1.9 (CoreCLR 4.700.20.47201, CoreFX 4.700.20.47203), X64 RyuJIT
  Job-AIVVBG : .NET Framework 4.8 (4.8.4250.0), X64 RyuJIT
  Job-KHFMMM : .NET Core 3.1.9 (CoreCLR 4.700.20.47201, CoreFX 4.700.20.47203), X64 RyuJIT
```

|           Method |        Job |       Runtime |     Toolchain |     Mean |   Error |  StdDev | Ratio |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------- |----------- |-------------- |-------------- |---------:|--------:|--------:|------:|-------:|------:|------:|----------:|
|    CurrentOnNext | Job-AIVVBG |    .NET 4.7.2 |        net472 | 185.0 ns | 1.43 ns | 1.19 ns |  1.00 | 0.0668 |     - |     - |     281 B |
|    CurrentOnNext | Job-KHFMMM | .NET Core 3.1 | netcoreapp3.1 | 183.4 ns | 0.76 ns | 0.63 ns |  0.99 | 0.0668 |     - |     - |     280 B |
|                  |            |               |               |          |         |         |       |        |       |       |           |
| CurrentOnNextNew | Job-AIVVBG |    .NET 4.7.2 |        net472 | 178.4 ns | 0.92 ns | 0.86 ns |  1.00 | 0.0668 |     - |     - |     281 B |
| CurrentOnNextNew | Job-KHFMMM | .NET Core 3.1 | netcoreapp3.1 | 174.6 ns | 0.94 ns | 0.88 ns |  0.98 | 0.0668 |     - |     - |     280 B |

There's still some improvement but not too drastic as before.